### PR TITLE
FileUpload Image Editor

### PR DESCRIPTION
**Pull request by tanthammar**
Add support for Cropper.js:
https://github.com/fengyuanchen/cropperjs

Via Fileponds official image edit plugin:
https://pqina.nl/filepond/docs/api/plugins/image-edit/

### I need help with:
- Make it look like Filament styling.
- Importing the cropperjs css into `packages/forms/resources/js/index.js`

Did translations for English/Swedish.
Can help with docs after all features are set.

The Croppable feature adheres to the FileUploads resize target width/height and aspect ratio.

Example:
```
SpatieMediaLibraryFileUpload::make('banner')
      ->imageResizeTargetWidth(1421)
      ->imageResizeTargetHeight(500)
      ->imageCropAspectRatio('16:9')
      ->croppable()
      ->cropperAspectRatios(['16:9', '4:3', '1:1', '2:3', null, '15:8'])
      ->collection('banner'),
```

dark-mode-cropper

cropper-light-mode

### CODE REVIEW

File: bin/build.mjs [hunk 1]
```diff
@@ -3,15 +3,18 @@ import * as esbuild from 'esbuild'
...
-
-    await context.rebuild()
-    await context.dispose()
+    const ctx = await esbuild.context(options)
+    if (isDev) {
+        await ctx.watch()
+    } else {
+        await ctx.rebuild()
+        await ctx.dispose()
+    }
```
Comment (zepfietje) [inline bin/build.mjs]: Please don't use abbreviations. 😄

File: package.json [hunk 2]
```diff
@@ -1,8 +1,8 @@
 {
     "name": "filament",
     "scripts": {
+        "dev:css": "npx tailwindcss -c ./packages/app/tailwind.config.js -i ./packages/app/resources/css/theme.css -o ./packages/app/dist/theme.css --minify --watch",
         "dev:js": "node bin/build.mjs --dev",
```
Comment (zepfietje) [inline package.json]: Please keep the order or also move the js scripts.

File: packages/forms/resources/js/components/file-upload.js [hunk 3]
```diff
@@ -7,6 +7,8 @@ import FilePondPluginImagePreview from 'filepond-plugin-image-preview'
 import FilePondPluginImageResize from 'filepond-plugin-image-resize'
 import FilePondPluginImageTransform from 'filepond-plugin-image-transform'
 import FilePondPluginMediaPreview from 'filepond-plugin-media-preview'
+import FilePondPluginImageEdit from 'filepond-plugin-image-edit'
+import Cropper from 'cropperjs'
```
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]: Alphabetical order.

File: packages/forms/resources/js/components/file-upload.js [hunk 4]
```diff
@@ -16,6 +18,7 @@ FilePond.registerPlugin(FilePondPluginImagePreview)
 FilePond.registerPlugin(FilePondPluginImageResize)
 FilePond.registerPlugin(FilePondPluginImageTransform)
 FilePond.registerPlugin(FilePondPluginMediaPreview)
+FilePond.registerPlugin(FilePondPluginImageEdit)
```
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]: Order.

File: packages/forms/resources/js/components/file-upload.js [hunk 5]
```diff
@@ -51,6 +54,14 @@ export default function fileUploadFormComponent({
...
     uploadProgressIndicatorPosition,
     uploadUsing,
+    disabled,
+    isCroppable,
+    cropperViewPortWidth,
+    cropperViewPortHeight,
+    shape,
+    updateCropperInputs,
+    viewMode,
+    fillColor,
```
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]: Order alpha.

File: packages/forms/resources/js/components/file-upload.js [hunk 6]
```diff
@@ -65,6 +76,14 @@ export default function fileUploadFormComponent({
 
         uploadedFileIndex: {},
 
+        showCropper: false,
+
+        editingFile: {},
+
+        currentRatio: '',
+
+        cropper: {},
```
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]: Same for these.

File: packages/forms/resources/js/components/file-upload.js [hunk 7]
```diff
@@ -261,9 +287,106 @@ export default function fileUploadFormComponent({
...
+        bindCropper(src) {
+            //avoid problems with cropper container not being visible when binding
+            setTimeout(() => {
+                this.cropper.replace(src)
+            }, 200)
         },
 
         destroy: function () {
+            this.cropper.destroy()
+            this.cropper = null
```
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]:
```suggestion
            this.cropper = null

```

File: packages/forms/resources/js/index.js [hunk 8]
```diff
@@ -6,6 +6,8 @@ import 'filepond/dist/filepond.min.css'
 import 'filepond-plugin-image-preview/dist/filepond-plugin-image-preview.css'
 import 'filepond-plugin-media-preview/dist/filepond-plugin-media-preview.css'
 import '../css/components/file-upload.css'
+import 'filepond-plugin-image-edit/dist/filepond-plugin-image-edit.css'
+import 'cropperjs/dist/cropper.min.css'
```
Comment (zepfietje) [inline packages/forms/resources/js/index.js]: Order.

File: packages/forms/resources/js/components/file-upload.js (hunk 7 above)
Comment (zepfietje) [inline packages/forms/resources/js/components/file-upload.js]:
```suggestion
            this.cropper.destroy()
            this.cropper = null

```

File: packages/forms/resources/lang/en/components.php [hunk 9]
```diff
@@ -297,4 +297,29 @@
...
+        'move-right' => 'Move image to right',
+        'move-up' => 'Move image up',
+        'move-down' => 'Move image down',
+        'move-top-left' => 'Move image to top/left corner of the container',
+        'rotate-left' => 'Rotate image to left',
+        'rotate-right' => 'Rotate image to right',
+        'rotate-180' => 'Rotate image 180°',
+        'flip-horizontal' => 'Flip image horizontal',
+        'flip-vertical' => 'Flip image vertical',
+    ],
```
Comment (zepfietje) [inline packages/forms/resources/lang/en/components.php]: Order.

File: packages/forms/resources/lang/sv/components.php [hunk 10]
```diff
@@ -271,5 +271,29 @@
...
+        'move-left' => 'Flytta bilden åt vänster',
+        'move-right' => 'Flytta bilden åt höger',
+        'move-up' => 'Flytta bilden uppåt',
+        'move-down' => 'Flytta bilden nedåt',
+        'move-top-left' => 'Flytta bilden till vänstra hörnet',
+        'rotate-left' => 'Rotera bilden åt vänster',
+        'rotate-right' => 'Rotera bilden åt höger',
+        'rotate-180' => 'Rotera bilden 180°',
+        'flip-horizontal' => 'Vänd bilden horisontellt',
+        'flip-vertical' => 'Vänd bilden vertikalt',
```
Comment (zepfietje) [inline packages/forms/resources/lang/sv/components.php]: Order.

File: packages/forms/resources/views/components/file-upload.blade.php [hunk 11]
```diff
@@ -8,6 +8,9 @@
         $imageResizeTargetHeight = $getImageResizeTargetHeight();
         $imageResizeTargetWidth = $getImageResizeTargetWidth();
         $statePath = $getStatePath();
+        $isDisabled = $isDisabled();
+        $cropperBtnClass = 'hover:bg-gray-600 dark:hover:bg-gray-800 hover:text-white focus:bg-gray-500 focus:ring-gray-500/50 flex-1 align-middle text-center font-medium outline-none transition-colors select-none focus:ring-2 border whitespace-no-wrap first-of-type:rounded-s-lg last-of-type:rounded-e-lg no-underline disabled:pointer-events-none disabled:opacity-70 filament-button-size-sm py-[calc(theme(spacing.2)-1px)] px-[calc(theme(spacing.[3.5])-1px)] text-sm shadow border-transparent';
+        $cropperActiveBtnClass = 'bg-gray-200 text-gray-800 dark:bg-gray-600 dark:text-white';
```
Comment (zepfietje) [inline packages/forms/resources/views/components/file-upload.blade.php]: Order.

**Review by zepfietje — CHANGES_REQUESTED** [review]
Left some quick initial comments, mostly around code style.

**Comment by tanthammar** [comment]
Hi @zepfietje I think you and Dan are supposed to adopt this and make it yours . So you are free to make all the changes you want :)

File: bin/build.mjs (hunk 1 above)
Comment (tanthammar) [reply bin/build.mjs]: I don't know what abbreviations mean ...
If it is "isDev" you mean, that is Dans creation :)
If it is "ctx", that is from esbuild documentation.

File: packages/forms/resources/views/components/file-upload.blade.php (hunk 11 above)
Comment (tanthammar) [reply packages/forms/resources/views/components/file-upload.blade.php]: ?

Comment (zepfietje) [reply packages/forms/resources/views/components/file-upload.blade.php]: Alphabetically.

**Comment by zepfietje** [comment]
We'll definitely do final tweaks, but if you could get some simple changes done, that'd save us time and allow us to review and merge the PR quicker.

In general, properties/variables are sorted alphabetically.
That's the main change I requested for now. 🙂

**Comment by tanthammar** [comment]
Tried to sort.

**Comment by awcodes** [comment]
@zepfietje I've updated this with similar styles to curators implantation. So unless Dan or Tina are still working on functionality I think it might be ready for review.

**Comment by zepfietje** [comment]
> @zepfietje I've updated this with similar styles to curators implantation. So unless Dan or Tina are still working on functionality I think it might be ready for review.

Okay thanks, Adam.
Could @tanthammar and @danharrin confirm this requires my review now?

**Review by danharrin — APPROVED** [review]

**Comment by danharrin** [comment]
@zepfietje ready for your review, I am happy

**Pull request merged**